Modules/_datetimemodule.c

```diff
@@ -1631,32 +1658,27 @@ wrap_strftime(PyObject *object, PyObject *format, PyObject *timetuple,
             ntoappend = 1;
         }
         /* A % has been seen and ch is the character after it. */
-        else if (ch == 'z') {
-            if (zreplacement == NULL) {
-                /* format utcoffset */
-                char buf[100];
-                PyObject *tzinfo = get_tzinfo_member(object);
-                zreplacement = PyBytes_FromStringAndSize("", 0);
-                if (zreplacement == NULL) goto Done;
-                if (tzinfo != Py_None && tzinfo != NULL) {
-                    assert(tzinfoarg != NULL);
-                    if (format_utcoffset(buf,
-                                         sizeof(buf),
-                                         "",
-                                         tzinfo,
-                                         tzinfoarg) < 0)
-                        goto Done;
-                    Py_DECREF(zreplacement);
-                    zreplacement =
-                      PyBytes_FromStringAndSize(buf,
-                                               strlen(buf));
-                    if (zreplacement == NULL)
-                        goto Done;
-                }
+        else if (ch == 'z' || (ch == ':' && *pin == 'z' && pin++)) {
+            /* %z -> +HHMM, %:z -> +HH:MM */
+            PyObject **replacement_p;
+            char *sep;
+            if (ch == ':') {
+                sep = ":";
+                replacement_p = &colonzreplacement;
+            } else {
+                sep = "";
+                replacement_p = &zreplacement;
+            }
+            if (*replacement_p == NULL) {
+                *replacement_p = make_somezreplacement(object,
+                                                       sep, tzinfoarg);
+                if (*replacement_p == NULL)
+                    goto Done;
             }
-            assert(zreplacement != NULL);
-            ptoappend = PyBytes_AS_STRING(zreplacement);
-            ntoappend = PyBytes_GET_SIZE(zreplacement);
+            assert(*replacement_p != NULL);
+            assert(PyBytes_Check(*replacement_p));
+            ptoappend = PyBytes_AS_STRING(*replacement_p);
+            ntoappend = PyBytes_GET_SIZE(*replacement_p);
```

Comment:
I still think that this could would gain in clarity if you gave up on the `replacement_p` variable and just did something like
```
if (ch == ':') {
    colonzreplacement = make_somezrelacement(object, ":", tzinfoarg);
    assert(...);
    assert(...);
    ptoappend = ...;
    ntoappend = ...;
}
else {
    // Do the same for %z
}

Reply:
hope it is ok like i did it in the last commit.